src/provisioningapi.ts

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2020 mx-puppet-bridge
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { PuppetBridge } from "./puppetbridge";
+import { Router, Request, Response } from "express";
+import { IPuppet } from "./db/puppetstore";
+
+const OK = 200;
+const CREATED = 201;
+const NO_CONTENT = 204;
+const BAD_REQUEST = 400;
+const UNAUTHORIZED = 401;
+const FORBIDDEN = 403;
+const NOT_IMPLEMENTED = 501;
+
+interface IPuppetWithDescription extends IPuppet {
+	description: string | null;
+}
+
+interface IAuthedRequest extends Request {
+	userId: string;
+}
+
+export class ProvisioningAPI {
+	private readonly apiRouterV1: Router;
+	private readonly apiSharedSecret: string;
+	constructor(
+		private readonly bridge: PuppetBridge,
+	) {
+		this.apiRouterV1 = Router();
+		this.apiSharedSecret = bridge.config.provisioning.sharedSecret;
+
+		const apiRouter = Router();
+		bridge.AS.expressAppInstance.use(bridge.config.provisioning.apiPrefix, apiRouter);
+		apiRouter.use(this.checkProvisioningSharedSecret);
+		apiRouter.use("/v1", this.apiRouterV1);
+
+		this.apiRouterV1.get("/status", this.status);
+		this.apiRouterV1.post("/link", this.link);
+		this.apiRouterV1.post("/:puppetId(\d+)/unlink", this.unlink);
+		this.apiRouterV1.get("/:puppetId(\d+)/users", this.listUsers);
+		this.apiRouterV1.get("/:puppetId(\d+)/rooms", this.listRooms);
+	}
+
+	get v1(): Router {
+		return this.apiRouterV1;
+	}
+
+	private async checkProvisioningSharedSecret(req: IAuthedRequest, res: Response, next: () => void) {
+		if (!this.apiSharedSecret) {
+			res.status(FORBIDDEN).json({
+				error: "The provisioning API is disabled",
+				errcode: "M_FORBIDDEN",
+			});
+		} else if (req.header("Authorization") !== `Bearer ${this.apiSharedSecret}`) {
+			res.status(UNAUTHORIZED).json({
+				error: "Unknown or missing token",
+				errcode: "M_UNKNOWN_TOKEN",
+			});
+		} else if (!req.query.user_id) {
+			res.status(BAD_REQUEST).json({
+				error: "Missing user_id query parameter",
+				errcode: "M_BAD_REQUEST",
+			});
+		} else {
+			req.userId = req.query.user_id;
+			next();
+		}
+	}
+
+	private async status(req: IAuthedRequest, res: Response) {
+		const puppets = await this.bridge.provisioner.getForMxid(req.userId) as IPuppetWithDescription[];
+		if (this.bridge.hooks.getDesc) {
+			for (const data of puppets) {
+				data.description = await this.bridge.hooks.getDesc(data.puppetId, data.data);
```

Comment:
perhaps add `delete data.data` as that object might contain rather sensitive information that someone using the API shouldn't be able to fetch?

Reply:
Isn't it always the integration manager or user themselves fetching the data?

Reply:
Soru is thinking of expanding this to e.g. have matrix-appservice-discord use it so that you can seemlessly start dms in mx-puppet-discord by inviting matrix-appservice-discord ghosts and it'd use that api, and you probably wouldn't want to allow access to that data, in that case.

Perhaps, to get back to the JWT idea, something with different secrets for different scopes? :thinking: